web/src/components/core/If.jsx

```diff
@@ -60,13 +60,15 @@
  *
  * @param {object} props
  * @param {boolean} props.condition
- * @param {JSX.Element} [props.then=null] - the content to be rendered when the condition is true
- * @param {JSX.Element} [props.else=null] - the content to be rendered when the condition is false
+ * @param {(JSX.Element|function():JSX.Element)} [props.then=null] - the content to be rendered when the condition is true
+ * @param {(JSX.Element|function():JSX.Element)} [props.else=null] - the content to be rendered when the condition is false
  */
 export default function If ({
   condition,
   then: positive = null,
   else: negative = null
 }) {
-  return condition ? positive : negative;
+  const ret = condition ? positive : negative;
+  // if the parameter is a function then evaluate it
+  return (typeof ret === "function") ? ret() : ret;
```

Comment:
This allows lazy evaluation of the branches. The problem is that `<If>` evaluates *both* branches and then returns one of them.

So for example this
```jsx
<If condition={data === undefined} then={<Loading/>} else={<Foo foo={data.foo}>} />
```

crashes when `data` is `undefined` because it evaluates the `else` branch too early.

With this improvement you can pass a function which can be evaluated only when the result is needed. (Passing functions is quite common in Javascript to allow the lazy evaluation.)

Then the fix is trivial, just add `() =>` before the value:
```jsx
<If condition={data === undefined} then={<Loading/>} else={() => <Foo foo={data.foo}>} />
```

@dgdavid are you OK with this change?

I'll add some example into the `<If>` documentation if you agree with this improvement.

Reply:
Thanks @lslezak 

Let's say that I have no strong opinion about it. At first sight, it looks reasonable to prevent not needed evaluations. But on the other hand it's weird to me to accept a function as content of the branches. IMHO it should be limited to [ReactNode](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/778ad7a80e9876f258270d84d1db4e6e1d5e3796/types/react/index.d.ts#L448-L489) (Although technically a component is a function too).


That said, I fully understand the proposal looking at the code written at https://github.com/openSUSE/agama/pull/1171/files#diff-994f1f4ff00fecec1797a2aaa799dbbeddc4e8fb2c1ae666fedbb95b7c499770R155-R178

```jsx
  <If
        condition={isLoading}
        then={
          <Loading text={_("Loading data...")} />
        }
        else={
          () =>
            <Form id="space-policy-form" onSubmit={onSubmit}>
              <SpacePolicyPicker currentPolicy={policy} onChange={setPolicy} />
              <If
                condition={devices.length > 0}
                then={
                  <SpaceActionsTable
                    devices={devices}
                    expandedDevices={expandedDevices}
                    deviceAction={deviceAction}
                    isActionDisabled={policy.id !== "custom"}
                    onActionChange={changeActions}
                  />
                }
              />
            </Form>
        }
      />
```

I would go for a different implementation instead. For me, what is in the else should be a component that React should be able to understand. Let's say `<SpacePolicyForm />`. Then, the component should be aware about what it needs for return either, an output or nothing.

```jsx
const SpacePolicyForm = ({ prop1, prop2 }) => {
  if (!prop1 || !prop2) return;

  ...

  return (
    <div><b>{prop1}:</b>{prop2}</div>
  );
}
```

I.e., the key for me is to write more robust components.

Of course, I might be wrong, so maybe another view here could help to decide what way to go.

Reply:
I was assuming the `else` branch is not evaluated if the condition is true. So, probably, I have written unsafe code.

Reply:
BTW, as far as I can see I think that this set of changes reveals another change we could address: to add the `isLoading` prop to core/Popup and make it show the spinner or its children based on the prop value. That way, we could reduce the `<If condition={isLoading} then={...` repetition and have more clean WhateverDialog.jsx components.

Reply:
After reading [this](https://medium.com/doctolib/the-subtle-aspects-of-react-re-rendering-part-1-functions-calls-and-lazy-evaluation-6136b0e31995) I decided to revert the `<If>` change. It seems it is not safe and it can have some nasty side effects in some cases.

Reply:
@joseivanlopez Actually I do not think you have written "unsafe" code. The `Form` itself is not executed (@dgdavid and me has done some tests with another example). The problem is the `data.foo` when `data` is undefined (that's the code that gets executed).